### ISSUE TITLE
Training time and testing demo

### ISSUE THREAD

**Issue by mjchen611** [wohlert/generative-query-network-pytorch#10]
Hi Jesper,

Thank you for your great code of gqn in real image,
I am a little curious about the following issues:
How many epochs it use to train a model on real image?
How many training data do you use (percentage of full training dataset)? 
Can you show a testing demo?

Thank you very much!

Best wishes,
Mingjia Chen

**Comment by wohlert** [comment]
Hi,

I trained the model on five part Shepard-Metzler dataset using only about 20% of the data. Training was run for 200,000 gradient steps instead of the 2,000,000 described in the paper. Training took about 14 hours on two NVIDIA Tesla-V100.

I cannot show a demo, because it seems like the GQN presented in the DeepMind paper was heavily overfitting to the dataset. If you want to achieve the same performance I suspect that you will need to use 100% of the dataset and run it for the full duration.

However, at the moment no one has been able to recreate their results.

**Comment by mjchen611** [comment]
Thank you for your answer, can you unpload an example of training data?
I use the data_convert code https://github.com/l3robot/gqn_datasets_translator
However, the code run_gqn.py always has an error in the part of data loading?

Thank you,
Best,
Mingjia

**Comment by wohlert** [comment]
After converting the dataset you should have a folder called `shepard_metzler_5_parts-torch/train` which should contain a list of PyTorch tensors, e.g. `0.pt`.

When running the `run_gqn.py` file you need to specify the location of this folder, for example.

`python3 run_gqn.py --data_dir "./shepard_metzler_5_parts-torch/train/"`

**Comment by mjchen611** [comment]
Thank you for your help,
it works now.

Mingjia

**Comment by versatran01** [comment]
@wohlert Just want to mention that I made a pull request to l3robot/gqn_datasets_translator which changed the format of the output file. It is possible that this will break your dataloader. See this discussion

https://github.com/l3robot/gqn_datasets_translator/issues/5

**Comment by mjchen611** [comment]
Hi @versatran01 
 
I read the code of l3robot/gqn_datasets_translator#5, I am a little curious that how long it will take to convert the full dataset shepard_metzler_5_parts?

Mingjia

**Comment by mjchen611** [comment]
Hi @wohlert 
can you upload an example file of training data,
the code run_gqn.py still has an error in the part of data loading?

Thank you,
Mingjia

**Comment by wohlert** [comment]
@versatran01 thanks for making the pull request to the translator! So far this has been my main bottleneck preventing me from training the model. I will change the `shepardmetzler.py` file to accomodate your new changes.

@mjchen611 when I have verified the new dataset updates and trained the model I will get back to you.

**Comment by versatran01** [comment]
@mjchen611 With the old one I couldn't finish 1% of the dataset for the entire night. 
For the new one, my rough estimate would be within 1 hour for my machine (32core) on ssd for sm5. YMMV.

**Comment by versatran01** [comment]
@wohlert The only problem with the new translator is that if you choose to convert the entire dataset, then each file would contain 2000 sequences, and one might need to split the data inside the training iteration. I was thinking of maybe dividing it to smaller chunks (eg. 200 seqs per file). What's your thought?

**Comment by wohlert** [comment]
Right now I have modified your code to output all the data in subfolders instead of gzipping it. So every file has the same number of records and have a structure like this for example `train/890-of-900/20.pt`.
It is still lightning fast to convert because of your multiprocessing code.

**Comment by versatran01** [comment]
Ok. Would love to know how long it takes in the end.

**Comment by mjchen611** [comment]
Hi @wohlert @versatran01 ,
thank you for your kind help, it works now.
I want to run this code on real image,  if possible can you give me an example code of the process (image convert to .tfrecord file,)?
thank you,
mingjia

**Comment by versatran01** [comment]
@mjchen611 

If you have real images, then you can just write a torch Dataset yourself, there's no need to convert it to tfrecord. The only reason tfrecord is used is because deepmind put their data in this format.

As for real images, from my understanding of the paper, gqn currently only works on synthetic toy datasets. Even the smallest dataset (sm5) is huge, containing 12 million images just for training. If you want to use it for real image, then you probably need to find a dataset of the same scale, and your network also needs much more capacity.  That's why deepmind didn't show any results of real images in their paper. Again just my thoughts, @wohlert might have more to say.

**Comment by mjchen611** [comment]
@versatran01 
thank you for your kind answer.

Mingjia

**Comment by mjchen611** [comment]
@wohlert how to use multi gpu?

**Comment by wohlert** [comment]
If you use the `run_gqn.py` script and set `--data_parallel=True` it should distribute the batch to all available GPUs.

**Comment by mjchen611** [comment]
Ok, thank your for your knid answer,
Best,
mingjia

**Comment by dthiem** [comment]
Hello @wohlert ,

first of all, great job you did here with an implementation.
Did you perhaps encountered some troubles regarding slow learning times during your work and have some tips on how to overcome those?

I tried to implement the article by myself, but i'm stuck with ridiculously long training times.
For 20% of shepard 5 parts set it takes up to 3 hours to do just one epoch on gtx 1070.
I do use data in gzips to save space, but to traverse 20% of set it takes only 2 minutes, wheras training GQN on same data takes up to 3 hours.

**Comment by wohlert** [comment]
Hi @dthiem 

I have not performed any performance analysis of the model, but I believe that the problem is just that the model specification makes it extremely computationally expensive. Jens Petersen has done an implementation based on this, in which he explains that [training took 22 days on a Titan Xp](https://github.com/jenspetersen/gqn-pytorch#speed-stability-and-performance).

I can try to provide some advice for speeding up training:
- set the number of generative layers lower, i.e. L=8
- use only a subset of the data

I would like to do an analysis of the run-time as I've found little to no advantage of multiple GPUs, but frankly I do not have the time at the moment.

**Comment by roveneliah** [comment]
Is there any way to download the trained weights directly?

**Comment by wohlert** [comment]
There are no trained weights currently.
